presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -942,6 +970,100 @@ public synchronized void rollback()
+        }));
+    }
+
+    private long getServerExpectedHeartbeatIntervalMillis()
+    {
+        String hiveServerTransactionTimeout = delegate.getConfigValue(HIVE_TXN_TIMEOUT.toString()).orElse("300s");
+        return toTime(hiveServerTransactionTimeout, getDefaultTimeUnit(HIVE_TXN_TIMEOUT), MILLISECONDS) / 2;
+    }
+
+    public synchronized Optional<ValidTxnWriteIdList> getValidWriteIds(ConnectorSession session, HiveTableHandle tableHandle)
+    {
+        String queryId = session.getQueryId();
+        checkState(currentQueryId.equals(Optional.of(queryId)), "Invalid query id %s while current query is", queryId, currentQueryId);
+        if (!tableHandle.isTransactional()) {
+            return Optional.empty();
+        }
+
+        checkState(hiveTransactionSupplier.isPresent(), "hiveTransactionSupplier is not set");
+        currentHiveTransaction = Optional.of(hiveTransactionSupplier.get().get());
+        return Optional.of(currentHiveTransaction.get().getValidWriteIds(delegate, tableHandle));
+    }
+
+    public synchronized void cleanupQuery(ConnectorSession session)
+    {
+        String queryId = session.getQueryId();
+        HiveIdentity identity = new HiveIdentity(session);
+        checkState(currentQueryId.equals(Optional.of(queryId)), "Invalid query id %s while current query is", queryId, currentQueryId);
+        Optional<HiveTransaction> transaction = currentHiveTransaction;
+        currentQueryId = Optional.empty();
+        currentHiveTransaction = Optional.empty();
+        hiveTransactionSupplier = Optional.empty();
+
+        if (!transaction.isPresent()) {
+            return;
+        }
+
+        long transactionId = transaction.get().getTransactionId();
+        ScheduledFuture<?> heartbeatTask = transaction.get().getHeartbeatTask();
+        boolean transactionValid = !heartbeatTask.isDone();
+        heartbeatTask.cancel(true);
+
+        // avoid race on HMS by waiting for heartbeat to complete
+        try {
+            heartbeatTask.get(5, SECONDS);
+        }
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+        }
+        catch (CancellationException | TimeoutException | ExecutionException ignored) {
+        }
+
+        if (transactionValid) {
+            // TODO if query fails, we should rollback (it doesn't matter for read-only transaction, but will be necessary in the future)
+            log.debug("Committing Hive transaction %s for query %s", transactionId, queryId);
+            delegate.commitTransaction(identity, transactionId);
+        }
+        else {
+            log.debug("Rolling back Hive transaction %s for query %s", transactionId, queryId);
+            delegate.rollbackTransaction(identity, transactionId);
+
+            throw new PrestoException(HIVE_TABLE_TRANSACTION_ABORTED_BY_METASTORE, format("Hive transaction %s was aborted by Metastore", transactionId));
```

Comment:
Successful commit would imply that data underneath hadnt changed as the Lock on that transaction would guarantee that. That guarantee is important in some cases of read-only transaction too e.g. with query reading same table in two sub-queries. AbortedTransaction or any other failure in commit may mean that two subqueries read different data and it would be better to fail the query in that case.

Reply:
If your question was on delegate.rollbackTransaction, I will be removing it as with the proposed changes of using commit irrespective of heartbeat status will make this redundant.